plugins/sfSkosPlugin/lib/sfSkosPlugin.class.php

```diff
@@ -17,359 +17,315 @@
+  /**
+   * We want to harvest the graph from top to bottom.
+   */
+  protected function getRootConcepts()
   {
-    $subjectUri = $concept->getAttributeNodeNS('http://www.w3.org/1999/02/22-rdf-syntax-ns#', 'about');
-    if (!($subjectUri instanceof DOMAttr) || !isset($this->terms[$subjectUri->nodeValue]))
+    $conceptScheme = $this->graph->get('skos:ConceptScheme', '^rdf:type');
+    if (null !== $conceptScheme)
     {
-      continue;
+      return $conceptScheme->allResources('skos:hasTopConcept');
     }
 
-    foreach ($this->xpath->query('./skos:related', $concept) as $related)
+    return $this->graph->allOfType('skos:Concept');
+  }
+
+  public function importGraph()
+  {
+    $concepts = $this->getRootConcepts();
+    foreach ($concepts as $item)
     {
-      $objectUri = $related->getAttributeNodeNS('http://www.w3.org/1999/02/22-rdf-syntax-ns#', 'resource');
-      if (!($objectUri instanceof DomAttr) || !isset($this->terms[$objectUri->nodeValue]))
+      if (false === $item instanceof EasyRdf_Resource)
       {
+        $this->logger->info($this->i18n->__('Unexpected concept, type received: %1%.', array('%1%' => gettype($item))));
+
         continue;
       }
 
-      // Don't duplicate relationship
-      foreach ($this->relations as $r)
+      $this->addConcept($item, $this->parent->id);
+    }
+
+    // Build list of relationships using sfSkosUniqueRelations
+    $relations = new sfSkosUniqueRelations;
+    foreach ($this->graph->allOfType('skos:Concept') as $c)
+    {
+      foreach ($c->allResources('skos:related') as $r)
       {
-        if (
-          $r['subject'] == $objectUri->nodeValue && $r['object'] == $subjectUri->nodeValue ||
-          $r['subject'] == $subjectUri->nodeValue && $r['object'] == $objectUri->nodeValue)
-        {
-          continue 2;
-        }
+        $relations->insert($c->get('atom:id')->getValue(), $r->get('atom:id')->getValue());
       }
+    }
 
+    foreach ($relations as $item)
+    {
       $relation = new QubitRelation;
       $relation->typeId = QubitTerm::TERM_RELATION_ASSOCIATIVE_ID;
-      $relation->subject = $this->terms[$subjectUri->nodeValue];
-      $relation->object = $this->terms[$objectUri->nodeValue];
+      $relation->subjectId = $item[0];
+      $relation->objectId = $item[1];
 
+      $relation->indexOnSave = false;
       $relation->save();
+    }
 
-      $this->relations[] = array('subject' => $subjectUri->nodeValue, 'object' => $objectUri->nodeValue);
+    // Report error with unsupported languages
+    if (0 < count($this->unsupportedLanguages))
+    {
+      $this->errors[] = $this->i18n->__('The following languages are used in the dataset imported but not supported by AtoM: %1%', array('%1%' => implode(",", array_keys($this->unsupportedLanguages))));
     }
 
-    return $this;
+    // Re-index parent term so numberOfDescendants reflects the changes
+    if (QubitTerm::ROOT_ID != $this->parent->id)
+    {
+      QubitSearch::getInstance()->update($this->parent);
```

Comment:
That'd be cool but where would you take the new value from? It'd be weird to instantiate a `arElasticSearchTermPdo` only for that reason. If I make the calculation then I'm duplicating code. If I use the ORM I'd need to hydrate the object again I believe?
It'd be cool if the `partialUpdate` method was able to generate the new value for me in case I don't give him once...